File: modules/gplazma2-banfile/src/test/scala/org/dcache/gplazma/plugins/BanFilePluginTest.scala
package org.dcache.gplazma.plugins

import org.scalatest._
import org.scalatest.junit.JUnitRunner
import org.junit.runner.RunWith

import scala.collection.JavaConversions._
import scala.io.Source

import java.util.Properties
import java.security.Principal

import org.dcache.gplazma.AuthenticationException
import com.google.common.io.Resources.getResource

class TestPrincipal(name : String) extends Principal {
  override def getName: String = name
  override def equals(other : Any) =
       other.getClass == this.getClass &&
       other.asInstanceOf[TestPrincipal].getName ==  name
}

// used in test "should not ban on equal names in different principal classes"
class TestPrincipalB(name : String) extends TestPrincipal(name)

trait stringSource extends BanFilePlugin {
  var sourceString : String = _
  override def fromSource : Source = Source fromString sourceString
}

@RunWith(classOf[JUnitRunner])
class BanFilePluginTest extends FlatSpec {

  def validProperties = {
    val properties = new Properties()
    properties.put(BanFilePlugin.BAN_FILE, getResource("ban.conf").getPath)
    properties
  }

  "The BanFilePlugin" should "throw an IllegalArgumentException if properties is null" in {
    intercept[IllegalArgumentException] {
      new BanFilePlugin(null)
    }
  }

  it should "throw an IllegalArgumentException if the gplazma.banfile.path property is not set" in {
    intercept[IllegalArgumentException] {
      new BanFilePlugin(new Properties)
    }
  }

  it should "throw an IllegalStateException if the config file does not exist at start time" in {
    val properties = new Properties()
    properties.put(BanFilePlugin.BAN_FILE, "/file/does/not/exist")

    intercept[IllegalStateException] {
      new BanFilePlugin(properties).account(null)
    }
  }

  it should "allow a not banned user" in {
    val plugin = new BanFilePlugin(validProperties) with stringSource
    plugin.sourceString =
      """
        |ban org.dcache.gplazma.plugins.TestPrincipal:bert
      """.stripMargin
    plugin.account(Set[Principal](new TestPrincipal("ernie")))
  }

  it should "not ban on equal names in different principal classes" in {
    val plugin = new BanFilePlugin(validProperties) with stringSource
    plugin.sourceString =
      """
        |ban org.dcache.gplazma.plugins.TestPrincipalB:ernie
      """.stripMargin
    plugin.account(Set[Principal](new TestPrincipal("ernie")))
  }

  it should "throw an AuthenticatedException for a banned user" in {
    val plugin = new BanFilePlugin(validProperties) with stringSource
    plugin.sourceString =
      """
        |ban org.dcache.gplazma.plugins.TestPrincipal:bert
      """.stripMargin
    intercept[AuthenticationException] {
      plugin.account(Set[Principal](new TestPrincipal("bert")))
    }
  }

  it should "work multiple times" in {
    val plugin = new BanFilePlugin(validProperties) with stringSource
    plugin.sourceString =
      """
        |ban org.dcache.gplazma.plugins.TestPrincipal:bert
      """.stripMargin
    plugin.account(Set[Principal](new TestPrincipal("ernie")))
    intercept[AuthenticationException] {
      plugin.account(Set[Principal](new TestPrincipal("bert")))
    }
    intercept[AuthenticationException] {
      plugin.account(Set[Principal](new TestPrincipal("bert")))
    }
    plugin.account(Set[Principal](new TestPrincipal("ernie")))
  }

  it should "ban user if any principal is blacklisted" in {
    val plugin = new BanFilePlugin(validProperties) with stringSource
    plugin.sourceString =
      """
        |ban org.dcache.gplazma.plugins.TestPrincipal:bert
      """.stripMargin
    intercept[AuthenticationException] {
      plugin.account(Set[Principal](new TestPrincipal("ernie"), new TestPrincipal("bert")))
    }
  }

  it should "use configured aliases" in {
    val plugin = new BanFilePlugin(validProperties) with stringSource
    plugin.sourceString =
      """
        |alias name=org.dcache.gplazma.plugins.TestPrincipal
        |ban name:bert
      """.stripMargin
    intercept[AuthenticationException] {
      plugin.account(Set[Principal](new TestPrincipal("bert")))
    }
  }

  it should "understand 'standard' principals without alias" in {
    val plugin = new BanFilePlugin(validProperties) with stringSource
    plugin.sourceString =
      """
        |ban name:bert
      """.stripMargin
    intercept[AuthenticationException] {
      plugin.account(Set[Principal](new org.dcache.auth.LoginNamePrincipal("bert")))
    }
  }

  it should "ignore comments" in {
    val plugin = new BanFilePlugin(validProperties) with stringSource
    plugin.sourceString =
      """
        |# some comment
        |alias name=org.dcache.gplazma.plugins.TestPrincipal
        |ban name:bert
      """.stripMargin
    intercept[AuthenticationException] {
      plugin.account(Set[Principal](new TestPrincipal("bert")))
    }
  }

  it should "ignore whitespace lines" in {
    val plugin = new BanFilePlugin(validProperties) with stringSource
    plugin.sourceString =
      """
        |alias name=org.dcache.gplazma.plugins.TestPrincipal
        |
        |ban name:bert
      """.stripMargin
    intercept[AuthenticationException] {
      plugin.account(Set[Principal](new TestPrincipal("bert")))
    }
  }

  it should "throw an IllegalArgumentException on malformed lines" in {
    val plugin = new BanFilePlugin(validProperties) with stringSource
    plugin.sourceString =
      """
        |this is not a valid line
      """.stripMargin
    intercept[IllegalArgumentException] {
      plugin.account(Set[Principal](new TestPrincipal("bert")))
    }
  }
}

File: modules/gplazma2-banfile/src/test/java/org/dcache/gplazma/plugins/BanFilePluginTest.java
package org.dcache.gplazma.plugins;

import org.dcache.auth.LoginNamePrincipal;
import org.dcache.auth.UserNamePrincipal;
import org.dcache.gplazma.AuthenticationException;
import org.junit.After;
import org.junit.Test;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.List;
import java.util.Properties;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

public class BanFilePluginTest {

    private BanFilePlugin plugin;
    private Path configFile;
    private AtomicInteger reloadCount = new AtomicInteger();

    @Test(expected = IllegalArgumentException.class)
    public void shouldFailWithNoProperties() {
        new BanFilePlugin(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void shouldFailWithMissingConfigProperty() {
        new BanFilePlugin(new Properties());
    }

    @Test(expected = IllegalArgumentException.class)
    public void shouldFailWithMissingConfigFile() {
        Properties properties = new Properties();
        properties.put(BanFilePlugin.BAN_FILE, "/some/missing/file");
        new BanFilePlugin(new Properties());
    }

    @Test
    public void shouldAllowNotBannedPrincipal() throws IOException, AuthenticationException {
        givenConfig("ban org.dcache.auth.UserNamePrincipal:bert");
        plugin.account(Set.of(new UserNamePrincipal("ernie")));
    }

    @Test
    public void shouldAllowSameNameDifferentClasses() throws IOException, AuthenticationException {
        givenConfig("ban org.dcache.auth.LoginNamePrincipal:bert");
        plugin.account(Set.of(new UserNamePrincipal("bert")));
    }

    @Test(expected = AuthenticationException.class)
    public void shouldFailForBannedUser() throws IOException, AuthenticationException {
        givenConfig("ban org.dcache.auth.UserNamePrincipal:bert");
        plugin.account(Set.of(new UserNamePrincipal("bert")));
    }

    @Test(expected = AuthenticationException.class)
    public void shouldFailIfAnyPrincipalIsBanned() throws IOException, AuthenticationException {
        givenConfig("ban org.dcache.auth.UserNamePrincipal:bert");
        plugin.account(Set.of(new LoginNamePrincipal("ernie"), new UserNamePrincipal("bert")));
    }

    @Test
    public void shouldWorkMultipleInvocations() throws IOException, AuthenticationException {
        givenConfig("ban org.dcache.auth.UserNamePrincipal:bert");
        try {
            plugin.account(Set.of(new UserNamePrincipal("bert")));
            fail();
        } catch (AuthenticationException expected) {
        }
        plugin.account(Set.of(new UserNamePrincipal("ernie")));

        try {
            plugin.account(Set.of(new UserNamePrincipal("bert")));
            fail();
        } catch (AuthenticationException expected) {
        }
        plugin.account(Set.of(new UserNamePrincipal("ernie")));
    }

    @Test(expected = AuthenticationException.class)
    public void shouldFailForBannedUserByAlias() throws IOException, AuthenticationException {
        givenConfig("alias foo=org.dcache.auth.UserNamePrincipal\nban foo:bert");
        plugin.account(Set.of(new UserNamePrincipal("bert")));
    }

    @Test(expected = AuthenticationException.class)
    public void shouldFailForBannedUserByStandardAlias() throws IOException, AuthenticationException {
        givenConfig("ban name:bert");
        plugin.account(Set.of(new LoginNamePrincipal("bert")));
    }

    @Test
    public void shouldIgnoreComments() throws IOException, AuthenticationException {
        givenConfig("# ban name:bert");
        plugin.account(Set.of(new LoginNamePrincipal("bert")));
    }

    @Test(expected = AuthenticationException.class)
    public void shouldAcceptFileWithComments() throws IOException, AuthenticationException {
        givenConfig("# this is a ban file\nban name:bert");
        plugin.account(Set.of(new LoginNamePrincipal("bert")));
    }

    @Test(expected = AuthenticationException.class)
    public void shouldFailForBannedUserByAliasAndBlanks() throws IOException, AuthenticationException {
        givenConfig("alias foo=org.dcache.auth.UserNamePrincipal\n   \n  ban foo:bert  ");
        plugin.account(Set.of(new UserNamePrincipal("bert")));
    }

    @Test(expected = IllegalArgumentException.class)
    public void shouldFailOnMalformedConfig() throws IOException, AuthenticationException {
        givenConfig("hello world!");
        plugin.account(Set.of(new UserNamePrincipal("bert")));
    }

    @Test(expected = IllegalArgumentException.class)
    public void shouldFailOnMalformedBanLine() throws IOException, AuthenticationException {
        givenConfig("ban foo");
        plugin.account(Set.of(new UserNamePrincipal("bert")));
    }

    @Test(expected = IllegalArgumentException.class)
    public void shouldFailOnMalformedAliasLine() throws IOException, AuthenticationException {
        givenConfig("alias foo");
        plugin.account(Set.of(new UserNamePrincipal("bert")));
    }

    @Test(expected = AuthenticationException.class)
    public void shouldBanAllIfConfigDeleted() throws IOException, AuthenticationException {
        givenConfig("");
        Files.deleteIfExists(configFile);
        plugin.account(Set.of(new UserNamePrincipal("bert")));
    }

    @Test
    public void shouldNotReReadConfigFileIfNotChanged() throws IOException, AuthenticationException, InterruptedException {
        givenConfig("ban org.dcache.auth.UserNamePrincipal:bert");
        plugin.account(Set.of(new UserNamePrincipal("ernie")));
        TimeUnit.SECONDS.sleep(1);
        plugin.account(Set.of(new UserNamePrincipal("ernie")));
        assertEquals("broken caching", 1, reloadCount.get());
    }

    @Test
    public void shouldReReadConfigFileIfChanged() throws IOException, AuthenticationException, InterruptedException {
        givenConfig("ban org.dcache.auth.UserNamePrincipal:bert");
        plugin.account(Set.of(new UserNamePrincipal("ernie")));

        TimeUnit.SECONDS.sleep(1);

        Files.writeString(configFile, "ban org.dcache.auth.LoginNamePrincipal:bert",
                StandardCharsets.UTF_8, StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.WRITE);
        plugin.account(Set.of(new UserNamePrincipal("ernie")));
        assertEquals("broken caching", 2, reloadCount.get());
    }

    private void givenConfig(String config) throws IOException {
        configFile = Files.createTempFile("dcache.gplazma.ban", "conf");
        Files.writeString(configFile, config, StandardCharsets.UTF_8, StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.WRITE);

        Properties properties = new Properties();
        properties.put(BanFilePlugin.BAN_FILE, configFile.toString());
        plugin = new BanFilePlugin(properties) {
            @Override
            List<String> loadConfigLines() throws IOException {
                reloadCount.incrementAndGet();
                return super.loadConfigLines();
            }
        };
    }

    @After
    public void tearDown() throws IOException {
        if (configFile != null) {
            Files.deleteIfExists(configFile);
        }
    }
}
